spec/controllers/scim_rails/scim_users_controller_spec.rb

```diff
@@ -0,0 +1,528 @@
+require "spec_helper"
+
+RSpec.describe ScimRails::ScimUsersController, type: :controller do
+  include AuthHelper
+
+  routes { ScimRails::Engine.routes }
+
+  describe "index" do
+    let(:company) { create(:company) }
+
+    context "when unauthorized" do
+      it "returns scim+json content type" do
+        get :index
+  
+        expect(response.content_type).to eq "application/scim+json, application/json"
+      end
+
+      it "fails with no credentials" do
+        get :index
+
+        expect(response.status).to eq 401
+      end
+
+      it "fails with invalid credentials" do
+        request.env['HTTP_AUTHORIZATION'] = ActionController::HttpAuthentication::Basic.encode_credentials("unauthorized","123456")
+
+        get :index
+
+        expect(response.status).to eq 401
+      end
+    end
+
+    context "when when authorized" do
+      before :each do
+        http_login(company)
+      end
+
+      it "returns scim+json content type" do
+        get :index
+  
+        expect(response.content_type).to eq "application/scim+json, application/json"
+      end
+
+      it "is successful with valid credentials" do
+        get :index
+
+        expect(response.status).to eq 200
+      end
+
+      it "returns all results" do
+        create_list(:user, 10, company: company)
+
+        get :index
+        response_body = JSON.parse(response.body)
+        expect(response_body.dig("schemas", 0)).to eq "urn:ietf:params:scim:api:messages:2.0:ListResponse"
+        expect(response_body["totalResults"]).to eq 10
+      end
+
+      it "defaults to 100 results" do
+        create_list(:user, 300, company: company)
+
+        get :index
+        response_body = JSON.parse(response.body)
+        expect(response_body["totalResults"]).to eq 300
+        expect(response_body["Resources"].count).to eq 100
+      end
+
+      it "paginates results" do
+        create_list(:user, 400, company: company)
+        expect(company.users.first.id).to eq 1
+
+        get :index, params: {
+          startIndex: 101,
+          count: 200,
+        }
+        response_body = JSON.parse(response.body)
+        expect(response_body["totalResults"]).to eq 400
+        expect(response_body["Resources"].count).to eq 200
+        expect(response_body.dig("Resources", 0, "id")).to eq 101
+      end
+
+      it "paginates results by configurable scim_users_list_order" do
+        ScimRails.config.scim_users_list_order = { created_at: :desc }
+
+        create_list(:user, 400, company: company)
+        expect(company.users.first.id).to eq 1
+
+        get :index, params: {
+          startIndex: 1,
+          count: 10,
+        }
+        response_body = JSON.parse(response.body)
+        expect(response_body["totalResults"]).to eq 400
+        expect(response_body["Resources"].count).to eq 10
+        expect(response_body.dig("Resources", 0, "id")).to eq 400
+
+        ScimRails.config.scim_users_list_order = nil
```

Comment:
I had two thoughts about this. One, I like the idea of stubbing the response so that the config is not actually modified (which in this case it was).

The second idea is that we should be freezing the values in the config file so that folks dont accidentally overwrite them from either code introduced in the Gem or code introduced in their apps. The problem with that is freezing every level of a hash or array and unintended consequences. I think leaving the config open is probably preferable to a hacky solution where we try and prevent changes to the config values outside the config file but I am open to suggestions.

Reply:
I don't think we should lock down the config. If someone wants to change it in runtime then that's on them.